Add Deployment Support to UnitedDeployment

uhari03 opened the pull request:
<!-- 
Please make sure you have read and understood the contributing guidelines;
https://github.com/openkruise/kruise/blob/master/CONTRIBUTING.md -->

### Ⅰ. Describe what this PR does
This PR adds Deployment Template support to the UnitedDeployment resource.

### Ⅱ. Does this pull request fix one issue?
<!--If so, add "fixes #xxxx" below in the next line, for example, fixes #15. Otherwise, add "NONE" -->
Fixes #262

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
Unit tests were done for deployment spec validation and for the uniteddeployment controller.

### Ⅳ. Describe how to verify it
Changes were included into the v0.6.1 helm chart.
Feel free to run `make manager` and `make docker-build` to build the docker image.
Then you can run `helm install` to install the kruise controller into your k8s cluster.

Here is an example of a UnitedDeployment YAML using deploymentTemplate:
```
apiVersion: apps.kruise.io/v1alpha1
kind: UnitedDeployment
metadata:
  labels:
    run: php-apache
  name: php-apache
  namespace: sam-system
spec:
  replicas: 4
  revisionHistoryLimit: 17
  selector:
    matchLabels:
      run: php-apache
  template:
    deploymentTemplate:
      metadata:
        labels:
          run: php-apache
      spec:
        selector:
          matchLabels:
            run: php-apache
        template:
          metadata:
            labels:
              run: php-apache
          spec:
            restartPolicy: Always
            dnsPolicy: ClusterFirst
            containers:
            - name: php-apache
              image: 614624372165.dkr.ecr.us-west-2.amazonaws.com/kruise-manager:hpa-example
              imagePullPolicy: Always
              ports:
              - containerPort: 80
                protocol: TCP
              resources:
                limits:
                  cpu: 500m
                requests:
                  cpu: 200m
              terminationMessagePolicy: File
  topology:
    subsets:
      - name: subset-a
        replicas: 50%
        nodeSelectorTerm:
          matchExpressions:
            - key: failure-domain.beta.kubernetes.io/zone
              operator: In
              values:
                - us-west-2a
      - name: subset-b
        replicas: 25%
        nodeSelectorTerm:
          matchExpressions:
            - key: failure-domain.beta.kubernetes.io/zone
              operator: In
              values:
                - us-west-2b
      - name: subset-c
        nodeSelectorTerm:
          matchExpressions:
            - key: failure-domain.beta.kubernetes.io/zone
              operator: In
              values:
                - us-west-2c

---
apiVersion: v1
kind: Service
metadata:
  name: php-apache
  namespace: sam-system
  labels:
    run: php-apache
spec:
  ports:
  - port: 80
  selector:
    run: php-apache
```

### Ⅴ. Special notes for reviews
- For easy reviewal purposes, updates have been made to v0.6.1 Helm charts. Not sure if this should instead involve a new minor version.
- Updates to documentation have not been included yet.
- Spec validation for Deployment objects is fairly minimal, but I think it is staying in-line with how statefulsets were validated.
- First time creating a PR to this repo, so please let me know of how I should improve :)

Fei-Guo commented on charts/kruise/v0.6.1/templates/apps.kruise.io_uniteddeployments.yaml:
```diff
@@ -752,6 +752,16 @@ spec:
                   required:
                   - spec
                   type: object
+                deploymentTemplate:
```
Let us not change this. The ud support of deployment will go live in next release.

Fei-Guo commented on api/openapi-spec/swagger.json:
```diff
@@ -0,0 +1,1243 @@
+{
```
We probably don't need this file.

Fei-Guo commented on charts/kruise/v0.6.1/templates/rbac_role.yaml:
```diff
@@ -193,6 +193,40 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
```
Don't change this file.

Fei-Guo commented on go.mod:
```diff
@@ -5,6 +5,7 @@ go 1.13
 require (
 	github.com/appscode/jsonpatch v1.0.1
 	github.com/codegangsta/negroni v1.0.0
+	github.com/go-openapi/spec v0.19.2
```
We don't need those packages if we remove the openapi spec from the PR.

Fei-Guo commented on pkg/controller/uniteddeployment/adapter/deployment_adapter.go:
```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2019 The Kruise Authors.
```
2019->2020

Fei-Guo commented on pkg/controller/uniteddeployment/adapter/deployment_adapter.go:
```diff
@@ -0,0 +1,259 @@
+	return nil
+}
+
+// PostUpdate does some works after subset updated. StatefulSets will implement this method to clean stuck pods.
```
Revise the comment since it is for Statefulsets.

Fei-Guo commented on pkg/controller/uniteddeployment/uniteddeployment_controller_statefulset_test.go:
```diff
@@ -1689,6 +1689,24 @@ func clean(g *gomega.GomegaWithT, c client.Client) {
 		return nil
 	}, timeout, time.Second).Should(gomega.Succeed())
 
+	deploymentList := &appsv1.DeploymentList{}
```
Why do you change this file? I think these tests are for statefulset subset only.

uhari03 replied:
I thought that this `clean` helper function did not have to be duplicated separately for Deployment objects.

The function seemed generic enough to add deployment support directly inside (since it already handles both statefulsets and advancedStatefulsets).

Let me know if you would still like me to separate it out.

Fei-Guo replied:
Ok. Fair enough. Then please add a comment for the `clean` method saying that it is shared by all subset workload tests. Please check cloneset test as well to see if it can use it.

uhari03 replied:
Seems like cloneset tests are already referencing this `clean` method.

Fei-Guo commented on pkg/controller/uniteddeployment/uniteddeployment_controller_deployment_test.go:
```diff
@@ -0,0 +1,1154 @@
+/*
+Copyright 2019 The Kruise Authors.
```
2019 -> 2020

Fei-Guo commented on pkg/webhook/uniteddeployment/validating/uniteddeployment_validation.go:
```diff
@@ -277,11 +295,16 @@ func validateAdvancedStatefulSet(statefulSet *appsv1alpha1.AdvancedStatefulSetTe
-		statefulSet.Spec.UpdateStrategy.RollingUpdate != nil &&
-		statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition != nil {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("spec", "updateStrategy", "rollingUpdate", "partition"), *statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition, "partition in advancedStatefulSetTemplate will not be used"))
+
```
Why is the original code removed?

uhari03 replied:
Hmm, might have been a merge error on my part, will fix it thanks

Fei-Guo reviewed (commented):
Thank for the change!  Comments are listed below.

FillZpp commented:
@uhari03 I have updated the travies ci today, please rebase the master branch and push again :)

LeoLiuYan commented:
ping @uhari03

uhari03 commented:
@LeoLiuYan @FillZpp @Fei-Guo 

Thanks for waiting on this, hopefully it wasn't too inconvenient

I hope I addressed all issues, please let me know if more changes are needed
